core/adapters/eth_tx.go

```diff
@@ -42,6 +44,85 @@ func (e *EthTx) TaskType() models.TaskType {
 // is not currently pending. Then it confirms the transaction was confirmed on
 // the blockchain.
 func (e *EthTx) Perform(input models.RunInput, store *strpkg.Store) models.RunOutput {
+	if store.Config.EnableBulletproofTxManager() {
+		return e.perform(input, store)
+	} else {
+		return e.legacyPerform(input, store)
+	}
+}
+
+// TODO(sam): Move away from resuming this on every new head and do something more sensible
+func (e *EthTx) perform(input models.RunInput, store *strpkg.Store) models.RunOutput {
+	trtx, err := store.FindEthTaskRunTxByTaskRunID(input.TaskRunID().UUID())
+	if err != nil {
+		logger.Error(err)
+		return models.NewRunOutputError(err)
+	}
+	if trtx != nil {
+		return e.checkForConfirmation(*trtx, input, store)
+	}
+	return e.insertEthTx(input, store)
+}
+
+func (e *EthTx) checkForConfirmation(trtx models.EthTaskRunTx, input models.RunInput, store *store.Store) models.RunOutput {
+	if trtx.EthTx.State == models.EthTxConfirmed {
+		receipt := models.EthReceipt{}
+		if err := store.GetRawDB().
+			Joins("INNER JOIN eth_tx_attempts ON eth_tx_attempts.hash = eth_receipts.tx_hash AND eth_tx_attempts.eth_tx_id = ?", trtx.EthTxID).
+			First(&receipt).Error; err != nil {
+			err = errors.Wrap(err, "could not find receipt for confirmed eth_tx")
+			logger.Error(err)
+			return models.NewRunOutputError(err)
+		}
+		output, err := models.JSON{}.Add("result", receipt.TxHash.Hex())
+		if err != nil {
+			logger.Error(err)
+			return models.NewRunOutputError(err)
+		}
+		return models.NewRunOutputComplete(output)
+	}
+	return models.NewRunOutputPendingOutgoingConfirmationsWithData(input.Data())
+}
+
+func (e *EthTx) insertEthTx(input models.RunInput, store *store.Store) models.RunOutput {
+	value, err := getTxData(e, input)
+	if err != nil {
+		err = errors.Wrap(err, "while constructing EthTx data")
+		return models.NewRunOutputError(err)
+	}
+
+	taskRunID := input.TaskRunID()
+	toAddress := e.ToAddress
+	var fromAddress common.Address
+	if e.FromAddress != nil {
+		fromAddress = *e.FromAddress
+	} else {
+		fromAddress, err = store.GetDefaultAddress()
+		if err != nil {
+			logger.Error(err)
+			return models.NewRunOutputError(err)
+		}
+	}
+	encodedPayload := utils.ConcatBytes(e.FunctionSelector.Bytes(), e.DataPrefix, value)
+
+	var gasLimit uint64
+	if e.GasLimit == nil {
+		gasLimit = store.Config.EthGasLimitDefault()
+	} else {
+		gasLimit = *e.GasLimit
+	}
+
+	if err := store.IdempotentInsertEthTaskRunTx(taskRunID, fromAddress, toAddress, encodedPayload, gasLimit); err != nil {
+		logger.Error(err)
```

Comment:
I'm noticing a pattern of handling errors with logging at multiple levels of the stack, I suspect when this code fails you'll get the same error being logged multiple times.

Reply:
You are correct we do log at multiple levels, and will get multiple errors.

I added the `logger.Error` at the lowest level since it helps debugging to see exactly where the error is coming from.

Any of these errors would be pretty serious so I don't see an issue in logging it multiple times... but if that is a problem, how would you prefer to solve the problem of logging being specific to the point of occurrence?

Reply:
We can and should be using Cheney's `errors` package which gets us stack traces, he also talks about how errors should only be logged once: https://dave.cheney.net/2016/04/27/dont-just-check-errors-handle-them-gracefully

Reply:
Ah, this looks like a much better solution. I'll use this errors package and wrap the errors at the bottom level, then we get a stacktrace at the top when they're logged out.